flink-test-utils-parent/flink-connector-test-utils/src/main/java/org/apache/flink/connector/testframe/utils/CollectIteratorAssert.java

```diff
@@ -0,0 +1,257 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.testframe.utils;
+
+import org.apache.flink.streaming.api.CheckpointingMode;
+
+import org.assertj.core.api.AbstractAssert;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+
+/** @param <T> */
+public class CollectIteratorAssert<T>
+        extends AbstractAssert<CollectIteratorAssert<T>, Iterator<T>> {
+
+    private static final int UNSET = -1;
+    private final Iterator<T> collectorIterator;
+    private final List<RecordsFromSplit<T>> recordsFromSplits = new ArrayList<>();
+    private int totalNumRecords;
+    private Integer limit = null;
+
+    protected CollectIteratorAssert(Iterator<T> collectorIterator) {
+        super(collectorIterator, CollectIteratorAssert.class);
+        this.collectorIterator = collectorIterator;
+    }
+
+    public CollectIteratorAssert<T> withNumRecordsLimit(int limit) {
+        this.limit = limit;
+        return this;
+    }
+
+    public void matchesRecordsFromSource(
+            List<List<T>> recordsBySplitsFromSource, CheckpointingMode semantic) {
+        for (List<T> recordsFromSplit : recordsBySplitsFromSource) {
+            recordsFromSplits.add(new RecordsFromSplit<>(recordsFromSplit));
+            totalNumRecords += recordsFromSplit.size();
+        }
+
+        if (limit != null && limit > totalNumRecords) {
+            throw new IllegalArgumentException(
+                    "Limit validation size should be less than total number of records from source");
+        }
+
+        switch (semantic) {
+            case AT_LEAST_ONCE:
+                matchAtLeastOnce(collectorIterator, recordsFromSplits);
+                break;
+            case EXACTLY_ONCE:
+                matchExactlyOnce(collectorIterator, recordsFromSplits);
+                break;
+            default:
+                throw new IllegalArgumentException(
+                        String.format("Unrecognized semantic \"%s\"", semantic));
+        }
+    }
+
+    private void matchAtLeastOnce(
+            Iterator<T> resultIterator, List<RecordsFromSplit<T>> recordsFromSplits) {
+        List<T> duplicateRead = new LinkedList<>();
```

Comment:
how about`duplicatedRecords` or   `recordsWithAtLeastOnceSemantic` ? And this method haven't unit test to cover in this PR?

Reply:
Fine, I will refactor this part.